### PR TITLE
utils: chunked_vector: various constructor improvements

### PR DESCRIPTION
**Pull request by avikivity**
Optimize the various constructors a little, and add an std::from_range_t
constructor.

Minor improvement, so no backports.

### CODE REVIEW

File: utils/chunked_vector.hh [hunk 1]
```diff
@@ -372,6 +377,28 @@ chunked_vector<T, max_contiguous_allocation>::chunked_vector(Iterator begin, Ite
...
+        size_t size = std::ranges::distance(range);
+        reserve(size);
+        auto begin = std::ranges::begin(range);
+        for (size_t i = 0; _size < size; ++i) {
+            T* dst = _chunks[i].get();
+            auto now = std::min(size - _size, max_chunk_capacity());
+            begin = std::ranges::uninitialized_copy_n(begin, now, dst, dst + now).in;
+            // Update _size incrementally to let the destructor
+            // know how much data to destroy on exception
+            _size += now;
```
Comment (nyh) [inline utils/chunked_vector.hh:395]: If I remember correctly, you had another copy of this exact same code. Would it make sense to create a (static) function that does it once?

**Review by nyh — APPROVED** [review]

File: utils/chunked_vector.hh (hunk 1 above)
Comment (avikivity) [reply utils/chunked_vector.hh:395]: It's not exactly the same.

**Comment by tchaikov** [comment]
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12909/) ✅ - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12909/) **The following new/updated tests ran 100 times for each mode:** 🔹 boost/chunked_vector_test ✅ - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/44/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12574/) ✅ - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1476/) ✅ - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4548/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12909/)
> #### Failed Tests (1/35679):
> 
>     * [[empty]](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/12909/testReport/junit/boost.secondary_index_test.test_secondary_index_contains_virtual_columns.1.xunit/xml/Tests___Unit_Tests____empty_) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+%5Bempty%5D)
> 

i *think* it is #20772

File: utils/chunked_vector.hh [hunk 2]
```diff
@@ -354,12 +354,20 @@ template <typename T, size_t max_contiguous_allocation>
...
+    constexpr auto is_forward = std::is_base_of<std::forward_iterator_tag, typename std::iterator_traits<Iterator>::iterator_category>::value;
+    if constexpr (is_forward) {
+        size_t size = std::distance(begin, end);
+        reserve(size);
+        for (size_t i = 0; _size < size; ++i) {
+            T* dst = _chunks[i].get();
+            auto now = std::min(size - _size, max_chunk_capacity());
+            begin = std::ranges::uninitialized_copy_n(begin, now, dst, dst + now).in;
+            // Update _size incrementally to let the destructor
+            // know how much data to destroy on exception
```
Comment (xemul) [inline utils/chunked_vector.hh:371]: What if above copy throws in the middle? The _size would be old, but some elements would already be there

Comment (avikivity) [reply utils/chunked_vector.hh:371]: The copy call will destroy elements it copied.

**Pull request merged**